db/view/row_locking.cc

```diff
@@ -184,8 +184,27 @@ row_locker::unlock(const dht::decorated_key* pk, bool partition_exclusive,
             lock.read_unlock();
         }
         if (!lock.locked()) {
+            auto& row_locks = pli->second._row_locks;
+            // verify that all rows are unlocked.
+            // a row might be held by lock_ck if the lock_partition future timed out
+            // but the lock_row future succeeded,
```

Comment:
So isn't that the real bug? If lock_ck()  got the ck lock but not the partition lock, it needs to clean up after itself and release the ck lock. Leaving the ck lock can create all sorts of bugs later - not just the unlock() error you're fixing here.

Reply:
What other sorts of bugs?
I don't think it's a bug by itself to leave stale row entries behind.
Erasing them should be an optimization.
We will not create them again if they are present.

In any case, a follow up series is cleaning this up so we won't leave stale entries on failure (#12187)

Reply:
It sounded to me like you described a case where lock_ck could still be locked (you used the word "held") forever without the partition being locked, which is wrong, not just wasteful. But I guess you meant something else - that an unlocked row remains in the array, not that its still lock.
Maybe you can update the commit message to have a better description of the case  you are solving? The issue has a lot of information but it's not clear to me how it relates to this actual fix.

Reply:
sure, I'll clarify that in the commit message and inline comments